### PR TITLE
pybind/mgr/status: add ceph pg osd-df-by-pool command to show pg distribution by pool

### PR DESCRIPTION
**Pull request by yanghonggang**
$ ceph pg osd-df-by-pool hehepool -f json
    {"osd_pg": [[0, 3], [1, 3], [2, 2], [3, 1], [4, 3]], "summary": {"MAX": 3, "AVG": 2.4, "MIN": 1}}
    
    $ ceph pg osd-df-by-pool hehepool
    +-----+-----+
    | OSD | PGS |
    +-----+-----+
    |  0  |  3  |
    |  1  |  3  |
    |  2  |  2  |
    |  3  |  1  |
    |  4  |  3  |
    +-----+-----+
    +-----+-----+-----+
    | MAX | MIN | AVG |
    +-----+-----+-----+
    |  3  |  1  | 2.4 |
    +-----+-----+-----+
    
    Signed-off-by: Yang Honggang <yanghonggang@umcloud.com>

### CODE REVIEW

File: src/mon/PGMap.cc [hunk 1]
```diff
@@ -3568,13 +3648,24 @@ int process_pg_map_command(
...
     pg_map.get_filtered_pg_stats(state, pool, osd, primary, pgs);
 
-    if (f && !pgs.empty()) {
-      pg_map.dump_filtered_pg_stats(f, pgs);
-      f->flush(*odata);
-    } else if (!pgs.empty()) {
-      pg_map.dump_filtered_pg_stats(ds, pgs);
-      odata->append(ds);
-      odata->append(omap_stats_note);
+    if (!osd_df_by_pool) {
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: i think new status-reporting commands should be implemented as mgr modules. so this command should be added to https://github.com/ceph/ceph/blob/e487604ad2af483a54e4616aff2f313a266d99d3/src/pybind/mgr/status/module.py#L17

@neha-ojha @liewegas what do you think?

Comment (yanghonggang) [reply src/mon/PGMap.cc]: @tchaikov  I implement this command in mgr modules.

File: src/pybind/mgr/status/module.py [hunk 2]
```diff
@@ -244,13 +252,74 @@ def handle_osd_status(self, cmd):
...
+        for p in pools:
+            if p['pool_name'] == pool:
+                poolid = p['pool']
+
+        if poolid is None:
+            return -errno.ENOENT, '', 'pool not found'
+
+        pgs = self.get_osdmap().map_pool_pgs_up(poolid)
+
+        osd2pgnums = {}
```
Comment (tchaikov) [inline src/pybind/mgr/status/module.py]:
```py
pgs_by_osd = defaultdict(int)
for osds in pgs.values():
  for osd in osd:
    pgs_by_osd[int(osd)] += 1
```

**Comment by yanghonggang** [comment]
> @yanghonggang after a second thought, i feel that this feature could be integrated as a part of balancer. where we prints number of pgs by osd, number of objects by osds, number of bytes by osds when evaluating the score of a cluster by its degree of balance from different perspective. what do you think?

@tchaikov 
OK, What's your suggestion about the new cmd's name?
`'balancer show-pg <pool>' ?`

**Comment by tchaikov** [comment]
@yanghonggang before moving forwards, i'd suggest we settle on a way to fit this feature into existing module.

@xiexingguo @liewegas what do you think?

**Comment by xiexingguo** [comment]
yeah, integrate the new command into balancer sounds a brilliant idea to me.

maybe `balancer show-pg-distribution <pool>`? if pool is unspecified, then fall back to print pg counters by osd.

**Comment by xiexingguo** [comment]
At first I thought it was weird that you still wanted to add a new pg stat command since we already had:

```
bin/ceph pg dump osds
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2019-09-17T18:46:55.674+0800 7fd995cbd700 -1 WARNING: all dangerous and experimental features are enabled.
2019-09-17T18:46:55.706+0800 7fd995cbd700 -1 WARNING: all dangerous and experimental features are enabled.
dumped osds
OSD_STAT USED    AVAIL    USED_RAW TOTAL   HB_PEERS  PG_SUM PRIMARY_PG_SUM 
4        1.8 GiB 1022 GiB  2.8 GiB 1.0 TiB [0,1,2,3]      3              2 
0        1.8 GiB 1022 GiB  2.8 GiB 1.0 TiB [1,2,3,4]      3              1 
1        1.8 GiB 1022 GiB  2.8 GiB 1.0 TiB [0,2,3,4]      3              1 
2        1.5 GiB 1022 GiB  2.5 GiB 1.0 TiB [0,1,3,4]      2              0 
3        1.3 GiB 1023 GiB  2.3 GiB 1.0 TiB [0,1,2,4]      1              0 
sum      8.3 GiB  5.0 TiB   13 GiB 5.0 TiB 
```

now I see the main difference is that you want a  per-pool (vs existing cluster-wide) output, which is only of significance for balancer for now, I guess?

**Comment by yanghonggang** [comment]
> At first I thought it was weird that you still wanted to add a new pg stat command since we already had:
> 
> ```
> bin/ceph pg dump osds
> *** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
> 2019-09-17T18:46:55.674+0800 7fd995cbd700 -1 WARNING: all dangerous and experimental features are enabled.
> 2019-09-17T18:46:55.706+0800 7fd995cbd700 -1 WARNING: all dangerous and experimental features are enabled.
> dumped osds
> OSD_STAT USED    AVAIL    USED_RAW TOTAL   HB_PEERS  PG_SUM PRIMARY_PG_SUM 
> 4        1.8 GiB 1022 GiB  2.8 GiB 1.0 TiB [0,1,2,3]      3              2 
> 0        1.8 GiB 1022 GiB  2.8 GiB 1.0 TiB [1,2,3,4]      3              1 
> 1        1.8 GiB 1022 GiB  2.8 GiB 1.0 TiB [0,2,3,4]      3              1 
> 2        1.5 GiB 1022 GiB  2.5 GiB 1.0 TiB [0,1,3,4]      2              0 
> 3        1.3 GiB 1023 GiB  2.3 GiB 1.0 TiB [0,1,2,4]      1              0 
> sum      8.3 GiB  5.0 TiB   13 GiB 5.0 TiB 
> ```
> 
> now I see the main difference is that you want a per-pool (vs existing cluster-wide) output, which is only of significance for balancer for now, I guess?

After checked ceph balancer's output, I find that all info I collected in this command already in ceph balancer eval-verbose's output.

```
 # ceph balancer eval-verbose hehepool
  ...
  count_by_pool {'hehepool': {'objects': {0: 0L, 1: 0L, 2: 0L, 3: 0L, 4: 0L}, 'bytes': {0: 0L, 1: 0L, 2: 0L, 3: 0L, 4: 0L}, 'pgs': {0: 3, 1: 3, 2: 2, 3: 1, 4: 3}}}
 ...
 stats_by_root {'default': {'objects': {'score': 0, 'avg': 0, 'sum_weight': 0, 'stddev': 0}, 'bytes': {'score': 0, 'avg': 0, 'sum_weight': 0, 'stddev': 0}, 'pgs': {'score': 0.11844759081950848, 'avg': 2.4, 'sum_weight': 0.6000000000000001, 'stddev': 0.8944271909999159}}
```

So  new version ceph don't need this command. I will close this pull request.  :)

**Pull request closed without merging**